src/proto/streams/streams.rs

```diff
@@ -687,18 +721,29 @@ where
 
         let me = &mut *me;
 
-        let id = {
-            let mut stream = me.store.resolve(self.key);
-            stream.ref_dec();
-
-            if !stream.is_released() {
-                return;
-            }
-
-            stream.remove()
-        };
+        let mut stream = me.store.resolve(self.key);
+        // decrement the stream's ref count by 1.
+        stream.ref_dec();
 
-        debug_assert!(!me.store.contains_id(&id));
+        let actions = &mut me.actions;
+        // the reset must be sent inside a `transition` block.
+        // `transition_after` will release the stream if it is
+        // released.
+        let recv_closed = stream.state.is_recv_closed();
```

Comment:
Oh hey, I missed this, I thought it was if the recv half is **NOT** closed...

Reply:
Fixed in 49627649ef6b2cea50bc81f228a21741a463636d

Reply:
So, `!recv_closed` makes some more sense, but I"m not sure I follow why we are waiting for the remote to close. If all handles are dropped, then we can't do anything w/ frames received from the remote. This is why we want to send a `RST_STREAM`, so that the remote can stop doing work?

Reply:
Yea, the point is to reset the stream if the user is no longer going to do anything with it. But, the the remote is already closed, there is no need to reset. At that point, it's just an unnecessary frame, that hopefully the remote will discard correctly, but it seems trivial for us to know right here whether the frame is needed...

Reply:
Ok, so what happens when the sender is still sending the body, but never sends an EOS and all the handles drop? At that point, a RST_STREAM should be sent out due to the sender aborting... I'm not sure where that happens.